File: package/python/src/sushi_seating.py
from typing import List
from collections import deque


def sushi_seating(commands: List[str]) -> List[str]:
    """
    回転寿司屋の席案内システムを実装する関数。

    Args:
        commands: 到着・席案内のコマンドリスト

    Returns:
        案内されたお客さんのリスト
    """
    waiting_queue: deque[str] = deque()  # FIFO queue for waiting customers
    seated_customers = []  # List of seated customers in order

    for command in commands:
        parts = command.split()

        if parts[0] == "arrive":
            name = parts[1]
            # Add to queue only if not already in queue
            if name not in waiting_queue:
                waiting_queue.append(name)

        elif parts[0] == "seat":
            n = int(parts[1])
            # Seat up to n customers from the front of the queue
            for _ in range(min(n, len(waiting_queue))):
                customer = waiting_queue.popleft()
                seated_customers.append(customer)

    return seated_customers

File: package/go/src/sushi_seating.go
package src

import (
	"strconv"
	"strings"
)

// SushiSeating implements a seat guidance system for a sushi restaurant.
//
// Parameters:
//   - commands: list of arrival and seating commands
//
// Returns:
//   - list of seated customers in order
func SushiSeating(commands []string) []string {
	waitingQueue := make([]string, 0)    // FIFO queue for waiting customers
	seatedCustomers := make([]string, 0) // List of seated customers in order

	for _, command := range commands {
		parts := strings.Split(command, " ")

		if len(parts) < 2 {
			continue
		}

		if parts[0] == "arrive" {
			name := parts[1]
			// Add to queue only if not already in queue
			if !contains(waitingQueue, name) {
				waitingQueue = append(waitingQueue, name)
			}

		} else if parts[0] == "seat" {
			n, err := strconv.Atoi(parts[1])
			if err != nil || n <= 0 {
				continue
			}
			// Seat up to n customers from the front of the queue
			seatedCount := minInt(n, len(waitingQueue))
			for i := 0; i < seatedCount; i++ {
				customer := waitingQueue[0]
				waitingQueue = waitingQueue[1:]
				seatedCustomers = append(seatedCustomers, customer)
			}
		}
	}

	return seatedCustomers
}

// contains checks if a string exists in a slice
func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

// minInt returns the smaller of two integers
func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}
